angr/simos/linux.py

```diff
@@ -306,3 +306,74 @@ def prepare_function_symbol(self, symbol_name, basic_addr=None):
             if basic_addr is None:
                 basic_addr = self.project.loader.extern_object.get_pseudo_addr(symbol_name)
             return basic_addr, basic_addr
+
+    def initialize_segment_register_x64(self, state, concrete_target):
```

Comment:
am I missing something? why aren't these methods on `concrete_target`?

Reply:
We chose this approach because we wanted to keep the target as simple as possible delegating the simos to deal with os/architecture specific issues. Moreover the simos or another angr component would still be required because, for example, on the x86 architecture we don't simply need to synchronize a register but we also have to store the gdt structure in memory.

Another approach we discussed about was adding to the target a `get_segment_register(os, arch)` method and let it deal with extracting the correct segment register depending on the architecture. Basically this means moving the logic of the following methods from the simos to the concrete target:
- `windows._read_fs_register_x86`
- `windows._read_gs_register_x64`
- `linux._read_gs_register_x86`
- `linux._read_fs_register_x64`

(yes the fs and gs depend on both the OS and the arch :-))

However we would still need methods in the simos like:
- `windows.initialize_segment_register_x64`
- `windows.initialize_gdt_x86`
- `linux.initialize_segment_register_x64`
- `linux.initialize_gdt_x86`

to set the correct registers or initialize the gdt depending on the os and architecture